zilliqa/src/api/eth.rs

```diff
@@ -816,6 +766,377 @@ fn get_uncle(_: Params, _: &Arc<Mutex<Node>>) -> Result<Option<String>> {
+        return Err(anyhow!("Too many topics provided, maximum 4"));
+    }
+
+    if node.config.max_filters > 0 && node.filters.len() as u64 > node.config.max_filters {
+        return Err(anyhow!("Too many filters already registered"));
+    }
+
+    let mut topics = [None, None, None, None];
+    topics[..params.topics.len()].swap_with_slice(&mut params.topics[..]);
+
+    let from_block = params.from_block.unwrap_or(BlockNumberOrTag::Latest);
+    // let first_block = node.resolve_block_number(from_block)?.unwrap().number();
+
+    let first_block = match params.from_block.unwrap_or(BlockNumberOrTag::Latest) {
+        BlockNumberOrTag::Number(n) => n,
+        tag => node.resolve_block_number(tag)?.unwrap().number(),
+    };
+
+    let mut rng = rand::thread_rng();
+
+    let mut id = [0u8; 32];
+    rng.fill(&mut id);
+
+    let id = B256::from_slice(&id);
+
+    node.filters.put(
+        id,
+        (
+            Filter::General(GeneralFilter {
+                from_block,
+                to_block: params.to_block.unwrap_or(BlockNumberOrTag::Latest),
+                address: params.address,
+                topics: Box::new(topics),
+            }),
+            first_block,
+            Instant::now(),
+        ),
+    );
+
+    Ok(id.to_hex())
+}
+
+fn uninstall_filter(params: Params, node: &Arc<Mutex<Node>>) -> Result<bool> {
+    let mut seq = params.sequence();
+    let filter_id: B256 = seq.next()?;
+    expect_end_of_params(&mut seq, 1, 1)?;
+
+    let mut node = node.lock().unwrap();
+
+    Ok(node.filters.pop(&filter_id).is_some())
+}
+
+fn new_block_filter(params: Params, node: &Arc<Mutex<Node>>) -> Result<String> {
+    expect_end_of_params(&mut params.sequence(), 0, 0)?;
+
+    let mut node = node.lock().unwrap();
+
+    if node.config.max_filters > 0 && node.filters.len() as u64 > node.config.max_filters {
+        return Err(anyhow!("Too many filters already registered"));
+    }
+
+    let first_block = node
+        .resolve_block_number(BlockNumberOrTag::Latest)?
+        .unwrap()
+        .number()
+        + 1;
+
+    let mut rng = rand::thread_rng();
+
+    let mut id = [0u8; 32];
+    rng.fill(&mut id);
+
+    let id = B256::from_slice(&id);
+
+    node.filters
+        .put(id, (Filter::Block, first_block, Instant::now()));
+
+    Ok(id.to_hex())
+}
+
+fn new_pending_transaction_filter(params: Params, node: &Arc<Mutex<Node>>) -> Result<String> {
+    expect_end_of_params(&mut params.sequence(), 0, 0)?;
+
+    let mut node = node.lock().unwrap();
+
+    if node.config.max_filters > 0 && node.filters.len() as u64 > node.config.max_filters {
+        return Err(anyhow!("Too many filters already registered"));
+    }
+
+    let first_block = node
+        .resolve_block_number(BlockNumberOrTag::Latest)?
+        .unwrap()
+        .number()
+        + 1;
+
+    let mut rng = rand::thread_rng();
+
+    let mut id = [0u8; 32];
+    rng.fill(&mut id);
+
+    let id = B256::from_slice(&id);
+
+    node.filters.put(
```

Comment:
Lots of the logic is duplicated between the 3 filter methods. Is it doable to de-duplicate at all?